File: 0221-maximal-square/0221-maximal-square.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    const m = matrix.length;
    const n = matrix[0].length;
    
    
    let dp = new Array(m);
    for(let r = 0; r < m;  r++) {
        dp[r] = new Array(n);
        dp[r].fill(0);
    }


    let max = 0;
    
    for(let r = 0; r < m;  r++) {
        for(let c = 0; c < n; c++) {
            if(matrix[r][c] == 1) {
                dp[r][c] = 1
                max = Math.max(max, dp[r][c])
            }
            if(r > 0 && c > 0) {
                if( matrix[r][c] == '1') {
                    dp[r][c] = 1 + Math.min(dp[r-1][c-1], Math.min(dp[r-1][c], dp[r][c-1]))
                }
                max = Math.max(max, dp[r][c])
            }
        }
    }
    //console.log(dp)
    
    
    return max * max
    
    
};

File: 0221-maximal-square/0221-maximal-square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        final int m = matrix.length;
        final int n = matrix.length > 0 ? matrix[0].length : 0;
        
        int[] prev = new int[n];
        int max = 0;
        
        for(int i = 0; i < m; i++) {
            int[] curr = new int[n];
            for(int j = 0; j < n; j++) {
                if(matrix[i][j] == '0') {
                    curr[j] = 0;
                } else {
                    curr[j] = 1;
                    if(i > 0 && j > 0)
                        curr[j] = Math.min( Math.min(prev[j-1], prev[j]), curr[j-1]) + 1;
                    max = Math.max(max,curr[j]);  
                }
            }
            prev = curr;
        }
        return max * max;
    }
}